sql/2020/07_SEO/core_web_vitals_by_device.sql

```diff
@@ -0,0 +1,75 @@
+#standardSQL
+# Core Web Vitals by device
+# Tests and notes at core_web_vitals_by_device.txt
+
+SELECT 
+device,
+
+# using adjusted values where fast+avg+slow add up to 1 (100%) for each row/metric group
+
+# Core Web Vitals indicate 75% good for all metrics is a pass
+ROUND(COUNTIF( fast_lcp_adjusted  > 0.75 AND fast_fid_adjusted  > 0.75 AND small_cls_adjusted  > 0.75) * 100 / COUNT(0), 2) AS corewebvitals_pass_pct,
+# pass based on each metric
+ROUND(COUNTIF( fast_lcp_adjusted  > 0.75) * 100 / COUNT(0), 2) AS pct_corewebvitals_pass_lcp,
+ROUND(COUNTIF( fast_fid_adjusted  > 0.75) * 100 / COUNT(0), 2) AS pct_corewebvitals_pass_fid,
+ROUND(COUNTIF( small_cls_adjusted  > 0.75) * 100 / COUNT(0), 2) AS pct_corewebvitals_pass_cls,
+
+# work out the grouped percents
+ROUND(AVG( fast_lcp_adjusted) * 100, 2) AS pct_fast_lcp,
```

Comment:
I'm writing a similar query for Performance chapter.
I understood [chrome UX report tips](https://developers.google.com/web/tools/chrome-user-experience-report#population-size) saying that aggregating density metrics across origins should be avoided.
These relative values are not linked to each other.

What do you expect the average value will show?

Reply:
Yes, I've been chatting about that with the SEO authors. 

In one sense we are wanting to find out average website performance, which I think means this sort of aggregation is not too bad. If we word it right. 

I guess the people in the Performance chapter would have a better grasp on those metrics. For SEO the Core Web Vitals will eventually become a ranking factor so we would like to say something about how websites are currently performing.

Reply:
@max-ostapenko @Tiggerito here's the best way to measure the percent of origins that pass the Core Web Vitals assessment in BigQuery. Let's all sync on this version for consistency.

```sql
CREATE TEMP FUNCTION IS_GOOD (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
  good / (good + needs_improvement + poor) >= 0.75
);

CREATE TEMP FUNCTION IS_NON_ZERO (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
  good + needs_improvement + poor > 0
);


SELECT
  # Origins with good LCP divided by origins with any LCP.
  SAFE_DIVIDE(
    COUNT(DISTINCT IF(IS_GOOD(fast_lcp, avg_lcp, slow_lcp), origin, NULL)),
    COUNT(DISTINCT IF(IS_NON_ZERO(fast_lcp, avg_lcp, slow_lcp), origin, NULL))) AS pct_good_lcp,
  
  # Origins with good FID divided by origins with any FID.
  SAFE_DIVIDE(
    COUNT(DISTINCT IF(IS_GOOD(fast_fid, avg_fid, slow_fid), origin, NULL)),
    COUNT(DISTINCT IF(IS_NON_ZERO(fast_fid, avg_fid, slow_fid), origin, NULL))) AS pct_good_fid,
  
  # Origins with good CLS divided by origins with any CLS.
  SAFE_DIVIDE(
    COUNT(DISTINCT IF(IS_GOOD(small_cls, medium_cls, large_cls), origin, NULL)),
    COUNT(DISTINCT IF(IS_NON_ZERO(small_cls, medium_cls, large_cls), origin, NULL))) AS pct_good_cls,
  
  # Origins with good LCP, FID, and CLS dividied by origins with any LCP, FID, and CLS.
  SAFE_DIVIDE(
    COUNT(DISTINCT IF(
      IS_GOOD(fast_lcp, avg_lcp, slow_lcp) AND
      IS_GOOD(fast_fid, avg_fid, slow_fid) AND
      IS_GOOD(small_cls, medium_cls, large_cls), origin, NULL)),
    COUNT(DISTINCT IF(
      IS_NON_ZERO(fast_lcp, avg_lcp, slow_lcp) AND
      IS_NON_ZERO(fast_fid, avg_fid, slow_fid) AND
      IS_NON_ZERO(small_cls, medium_cls, large_cls), origin, NULL))) AS pct_good_cwv
FROM
  `chrome-ux-report.materialized.metrics_summary`
WHERE
  date = '2020-06-01'
```

Reply:
@max-ostapenko @rviscomi Should we have the official queries in the Performance chapter. I'm happy to beg/borrow/steal them 😃

Reply:
@rviscomi We need metrics segmented by geo/devices/network, so I'm using `chrome-ux-report.experimental.country` table for that purpose instead.
And I was doing all metrics calculation combined in one query (it's quite an expensive query), as opposed to running a separate query per chart. Let me know if you see the disadvantages of this approach.

@Tiggerito sure, let's see if our chapter provides all Performance data required. Here is our draft PR: #1091

Reply:
And thanks for this example query, I found I was considering NULL metrics as poor category - fixing.

Reply:
> @rviscomi We need metrics segmented by geo/devices/network, so I'm using `chrome-ux-report.experimental.country` table for that purpose instead.
> And I was doing all metrics calculation combined in one query (it's quite an expensive query), as opposed to running a separate query per chart. Let me know if you see the disadvantages of this approach.
> 
> @Tiggerito sure, let's see if our chapter provides all Performance data required. Here is our draft PR: #1091

We just need country and device, so that would cover us :-)